File: javascript/13.roman-to-integer.js
/*
 * @lc app=leetcode id=13 lang=javascript
 *
 * [13] Roman to Integer
 */

// @lc code=start
/**
 * @param {string} s
 * @return {number}
 */
var romanToInt = function (s) {
  const symbols = {
    I: 1,
    V: 5,
    X: 10,
    L: 50,
    C: 100,
    D: 500,
    M: 1000,
  };
  let result = 0;
  let i = 0;
  for (i; i < s.length - 1; i++) {
    const p = symbols[s[i]],
      q = symbols[s[i + 1]];
    if (p < q) result -= p;
    else result += p;
  }
  result += symbols[s[i]];
  return result;
};
// @lc code=end

File: rust/13.roman-to-integer.rs
/*
 * @lc app=leetcode id=13 lang=rust
 *
 * [13] Roman to Integer
 */

// @lc code=start
impl Solution {
    pub fn roman_to_int(s: String) -> i32 {
        use std::collections::HashMap;

        let map = |c: &char| match c {
            'I' => 1,
            'V' => 5,
            'X' => 10,
            'L' => 50,
            'C' => 100,
            'D' => 500,
            'M' => 1000,
            _ => 0,
        };

        let mut p = '\0';
        let mut result = map(&p);
        for c in s.chars().rev() {
            let value = map(&c);
            result += match map(&p) > value {
                true => -value,
                false => value,
            };
            p = c;
        }
        return result;
    }
}
// @lc code=end
